content/docs/rfcs/35/README.md

```diff
@@ -0,0 +1,143 @@
+
+Four Noise handshakes are currently supported: `K1K1`, `XK1`, `XX`, `XXpsk0`. Their descriptions can be found [here](https://forum.vac.dev/t/noise-handshakes-as-key-exchange-mechanism-for-waku2/130). 
+
+We note that all design requirements on exchanged messages would be satisfied only *after* a supported Noise handshake is completed, all corresponding to a total of 1 Round Trip Time communication *(1-RTT)*.
+
+In the following, we assume that communicating parties reciprocally know an initial [`contentTopic`](https://rfc.vac.dev/spec/14/#wakumessage) where they can send/receive their first handshake messages. 
+
+## Cryptographic primitives
+
+Noise protocols are instantiated with the following underlying cryptographic primitives:
+- `Curve25519` for Diffie-Hellman key-exchanges;
+- `ChaChaPoly` for symmetric authenticated encryption;
+- `SHA256` used in `HMAC` and `HKDF` keys derivation chains.
+
+
+## Specification
+
+For 10/WAKU2, the `payload` field is used in `WakuMessage` and MAY contain a Noise handhshake message and a Noise transport message.
+
+The fields that are concatenated as part of the `payload` Waku2 field are:
+
+ - `protocol-id`: identifies the protocol in use (1 byte). 
+ Supported values are:
+     - `0`: none or omitted;
+     - `10`: Noise protocol `Noise_K1K1_25519_ChaChaPoly_SHA256`;
+     - `11`: Noise protocol `Noise_XK1_25519_ChaChaPoly_SHA256`;
+     - `12`: Noise protocol `Noise_XX_25519_ChaChaPoly_SHA256`;
+     - `13`: Noise protocol `Noise_XXpsk0_25519_ChaChaPoly_SHA256`;
+ - `handshake-length`: the length in bytes of the Noise handshake message (1 byte);
+ - `handshake-message`: the Noise handshake message. It contains the concatenation of one or more Diffie-Hellman public keys `pk` (ephemeral or static).
+A public key `pk` is encoded as `(flag || pk_serialized || pk_ad)`, where:
+     - `flag`: is equal to `1` if the `pk_serialized` field is encrypted; is equal to `0` if `pk_serialized` is not encrypted (1 byte);
+     - `pk_encoding`: a `Curve25519` serialization of the public key. If `flag = 1`, this field is encrypted (56 bytes);
+     - `pk_ad`: if `flag = 1`, it contains the authentication data for the encrypted field `pk_encoding` (16 bytes). It is empty otherwise (0 bytes).
+ - `message-len-size`: the number of bytes required to represent the Noise transport message length in bytes (1 byte);
+ - `message-length`: the length in bytes of the Noise transport message (`message-len-size` bytes);
```

Comment:
It is used to overall reduce data. Since, in principle, the `transport message` can have arbitrary size, we can store its size in either two way: 
- a constant-sized field `message-length`;
- using two fields `message-len-size` and `message-length`.

In the first case, `message-length` should be at least 4 bytes to accommodate at least 4TB size for `transport message`, but short messages (the most common ones) will always set its MSB to 0.

In the second case, `message-len-size`, which is 1 byte, will tell how many bytes `message-length` is, so that for short messages you use only 2-3 bytes, rather than 4. This virtually allows `message-length` to be up to 256 bytes.

In practice, since I don't think we need more than 8 bytes for  `message-length`, we can reserve only 3 bits for `message-len-size`. To not use a single byte as in current commit, these 3 bits can be eventually embedded in the `protocol-id` field.